temporalio/client.py

```diff
@@ -5298,50 +5924,117 @@ async def start_workflow_update(
                 else:
+                    multiop_failure = (
+                        temporalio.api.errordetails.v1.MultiOperationExecutionFailure()
+                    )
+                    if err.grpc_status.details[0].Unpack(multiop_failure):
+                        status = next(
+                            (
+                                st
+                                for st in multiop_failure.statuses
+                                if (
+                                    st.details
+                                    and not st.details[0].Is(
+                                        temporalio.api.failure.v1.MultiOperationExecutionAborted.DESCRIPTOR
+                                    )
+                                )
+                            ),
+                            None,
+                        )
+                        if status and status.code in RPCStatusCode:
+                            raise RPCError(
+                                status.message,
+                                RPCStatusCode(status.code),
+                                err.raw_grpc_status,
+                            )
                     raise
-            except asyncio.CancelledError as err:
-                raise WorkflowUpdateRPCTimeoutOrCancelledError() from err
+
+            start_responses: List[
+                temporalio.api.workflowservice.v1.StartWorkflowExecutionResponse
+            ] = [
+                r.start_workflow
+                for r in multiop_response.responses
+                if r.HasField("start_workflow")
+            ]
+            update_responses: List[
+                temporalio.api.workflowservice.v1.UpdateWorkflowExecutionResponse
+            ] = [
+                r.update_workflow
+                for r in multiop_response.responses
+                if r.HasField("update_workflow")
+            ]
+            if not (len(start_responses) == 1 and len(update_responses) == 1):
+                raise RuntimeError("Invalid ExecuteMultiOperationResponse")
+            [start_response] = start_responses
+            [update_response] = update_responses
+            if not seen_start:
+                input._on_start_callback(start_response)
+                seen_start = True
+            known_outcome = (
+                update_response.outcome if update_response.HasField("outcome") else None
+            )
             if (
-                resp.stage >= req.wait_policy.lifecycle_stage
-                or resp.stage
+                update_response.stage
                 >= temporalio.api.enums.v1.UpdateWorkflowExecutionLifecycleStage.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED
             ):
                 break
 
-        # Build the handle. If the user's wait stage is COMPLETED, make sure we
-        # poll for result.
-        handle: WorkflowUpdateHandle[Any] = WorkflowUpdateHandle(
+        return WorkflowUpdateHandle(
```

Comment:
When we do `execute_update` (or the user set wait stage to completed), we don't return the handle until we have polled for outcome

Reply:
You're right, there's a test that I'd intended to be confirming that the result is fetched in that context, but the test was misconceived. It's not the easiest thing to test for, but I've put the poll call in. Maybe we can add a test later based on manipulating the history long poll timeout.

```python
                break

        handle = WorkflowUpdateHandle(
            client=self._client,
            id=update_req.request.meta.update_id,
            workflow_id=start_input.id,
            workflow_run_id=start_response.run_id,
            known_outcome=known_outcome,
        )
        if update_input.wait_for_stage == WorkflowUpdateStage.COMPLETED:
            await handle._poll_until_outcome()

        return handle
```